Topics/Development_Process/GraphQL_VS_REST.md

```diff
@@ -0,0 +1,94 @@
+
+On this page, we will refer to things called `endpoints`. These are basically access ways for our API - you use these endpoints to communicate with the software.
+
+### Why do we need APIs?
+Without APIs, there would be no way for software to communicate between one another. In fact, even within the same project, different components may have APIs between them to facilitate communication. Without APIs, software would never be able to interact with anything. Even something as simple as loading Google's webpage deals with many dozens of APIs without you knowing! 
+
+### Designing APIs
+In this article, we will specifically discuss APIs that deal with sending requests over the internet, which is a very special case of APIs. We will cover two special commonly used types of APIs:
+1. REST
+2. GraphQL 
+
+On this page, you will learn about the differences between the two, and hopefully be able to make a confident decision on which one you need to use, based on your project requirements. 
+
+## REST Overview
+A REST API is any API that follows the REST architecture. The implementation of it can be handled in many pre-existing libraries, so this page will focus on the important details:
+1. In REST, you have to specify the `type` of the request you are sending when you call an API endpoint.
+    + Some common examples include `GET`, `POST`, `PUT`, and `DELETE`. Others exist, but these are the ones used most often.
+2. REST is not specific; when you call an endpoint, you will get *all* of the data that endpoint can give, whether you like it or not.
+    + This can be ideal when you require a lot of data, but can also lead to larger than necessary load on your system if you have to call many API endpoints in quick succession.
+3. Every difference type of request in REST has to be called to a different endpoint. For instance, if you want to `GET` all `Users` and `Cars` at the same time, with a typical REST API, you would need to send two seprate `GET` requests to a `/users` and `/cars` endpoint.
+
+
+## GraphQL Overview
+GraphQL is difference from REST in the sense that it is not actually an API architecture. In fact, it is actually a `query language`, hence the QL in its name. This leads to a number of distinct characteristics that we will discuss:
+1. In GraphQL there are only two types of requests: `Queries`, and `Mutations`. A query is used to get data, i.e, "query" some data. Everything else would be a mutation; creating, updating, or deleting data all mutate, and thus they are mutations.
+    + To follow proper GraphQL standards, any endpoint that only sends back data would fall under `query`. Any endpoint that changes data in any way would fall under `mutation`.
+2. GraphQL is very specific; when you call an endpoint, you can request *exactly* what data you want, and you can send different structures of data, based on customizable `types`.

```

Comment:
I'm not sure how this would be different from REST other than only having 1 endpoint. Can't you also request exactly what data you want in REST?

Reply:
I think we may have already resolved this in a previous conversation; could you confirm that? Thanks!

Reply:
Great!